lib/hammer_cli/defaults.rb

```diff
@@ -0,0 +1,207 @@
+    end
+
+    def self.get_defaults
+      raise NotImplementedError
+    end
+  end
+
+  class DefaultsCommand < HammerCLI::AbstractCommand
+
+    class ListDefaultsCommand < HammerCLI::DefaultsCommand
+      command_name 'list'
+
+      desc _('List all the default parameters')
+      def execute
+        list_all_defaults_message
+        HammerCLI::EX_OK
+      end
+    end
+
+    class DeleteDefaultsCommand < HammerCLI::DefaultsCommand
+      command_name 'delete'
+
+      desc _('Delete a default param')
+      option "--param-name", "OPTION_NAME", _("The name of the default option"), :required => true
+
+      def execute
+        if context[:defaults].defaults_settings && context[:defaults].defaults_settings[option_param_name.to_sym]
+          context[:defaults].delete_default_from_conf(option_param_name.to_sym)
+          param_deleted(option_param_name)
+        else
+          variable_not_found
+        end
+        HammerCLI::EX_OK
+      end
+    end
+
+    class AddDefaultsCommand < HammerCLI::DefaultsCommand
+      command_name 'add'
+
+      desc _('Add a default parameter to config')
+      option "--param-name", "OPTION_NAME", _("The name of the default option"), :required => true
+      option "--param-val", "OPTION_VALUE", _("The value for the default option")
+      option "--plugin-name", "OPTION_PLUGIN_NAME", _("The plugin name defaults will be generated for")
+
+      def execute
+        if option_plugin_name.nil? && option_param_val.nil? || !option_plugin_name.nil? && !option_param_val.nil?
+          bad_input
+          HammerCLI::EX_USAGE
+        else
+          if option_plugin_name
+            namespace = option_plugin_name
+            if !context[:defaults].providers.key?(namespace)
+              plugin_prob_message
+              return HammerCLI::EX_USAGE
+            elsif !context[:defaults].providers[namespace].support?(option_param_name)
+              defaults_not_supproted_by_plugin
+              return HammerCLI::EX_CONFIG
+            end
+          end
+          context[:defaults].add_defaults_to_conf({option_param_name => option_param_val}, namespace)
+          added_default_message(option_param_name.to_s, option_param_val)
+          HammerCLI::EX_OK
+        end
+      rescue Defaults::DefaultsError, SystemCallError => e
+        print_message(e.message)
+        HammerCLI::EX_CONFIG
+      end
+    end
+
+    def added_default_message(key, value)
+      print_message(_("Added %{key_val} default-option with value that will be generated from the server.") % {:key_val => key.to_s} ) if value.nil?
+      print_message(_("Added %{key_val} default-option with value %{val_val}.") % {:key_val => key.to_s, :val_val=> value.to_s} ) unless value.nil?
+    end
+
+    def plugin_prob_message
+      print_message(_("Couldn't reach the plugin defaults class, eiter the plugin doesn't have defaults class or the plugin misspelled.(example, for Hammer_CLI_Foreman --plugin-name foreman)"))
+    end
```

Comment:
This seems quite difficult to grasp at a first glance. Can it say "Unknown provider. To list available providers run `hammer defaults providers`" and `default providers` would print list of registered providers and supported defaults?

I also see mix of plugin and provider in the messages. It would be simplier to have just defaults providers registered under some name and not imposing implicit connection to plugin that implements the provider.